OpenCV 4 fails to build with cuda enabled due to missing NVIDIA optical flow sdk

**Describe the bug**
OpenCV 4 fails to build with cuda enabled due to missing NVIDIA optical flow sdk

```
In file included from /build/source/modules/core/include/opencv2/core.hpp:54,
                 from /build/source/modules/core/include/opencv2/core/cuda.hpp:51,
                 from /build/source/opencv_contrib/cudaoptflow/include/opencv2/cudaoptflow.hpp:50,
                 from /build/source/opencv_contrib/cudaoptflow/src/precomp.hpp:48,
                 from /build/source/opencv_contrib/cudaoptflow/src/nvidiaOpticalFlow.cpp:7:
/build/source/modules/core/include/opencv2/core/base.hpp:320:40: error: expected constructor, destructor, or type conversion before '(' token
  320 | #define CV_Error( code, msg ) cv::error( code, msg, CV_Func, __FILE__, __LINE__ )
      |                                        ^
/build/source/opencv_contrib/cudaoptflow/src/nvidiaOpticalFlow.cpp:15:1: note: in expansion of macro 'CV_Error'
   15 | CV_Error(cv::Error::HeaderIsNull, "Nvidia Optical Flow headers not found. Make sure cmake downloads it properly");
      | ^~~~~~~~
make[2]: *** [modules/cudaoptflow/CMakeFiles/opencv_cudaoptflow.dir/build.make:110: modules/cudaoptflow/CMakeFiles/opencv_cudaoptflow.dir/src/nvidiaOpticalFlow.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:5658: modules/cudaoptflow/CMakeFiles/opencv_cudaoptflow.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
builder for '/nix/store/jiklcp5gzafrszikvrglykqiffk873js-opencv-4.1.2.drv' failed with exit code 2
error: build of '/nix/store/jiklcp5gzafrszikvrglykqiffk873js-opencv-4.1.2.drv' failed
```

**To Reproduce**
Steps to reproduce the behavior:
1. nix-build -A opencv4

**Expected behavior**
Either we disable OpenCV 4's option to compile with the NVIDIA optical flow SDK or allow compilation against the NVIDIA optical flow sdk headers.

See: https://devtalk.nvidia.com/default/topic/1066111/cuda-setup-and-installation/nvidiaopticalflow-header-missing-opencv/

I've done some preliminary work towards this on the following branch:
https://github.com/NixOS/nixpkgs/commit/fda377a880e8eebd92010d7fe07f5685e00f184b

--- a/pkgs/development/libraries/nvidia-optical-flow-sdk/default.nix
+++ b/pkgs/development/libraries/nvidia-optical-flow-sdk/default.nix
@@ -1,0 +1,28 @@
+{ stdenv, fetchFromGitHub }:
+
+stdenv.mkDerivation {
+  pname = "nvidia-optical-flow-sdk";
+  version = "1.0";
+
+  src = fetchFromGitHub {
+    owner = "NVIDIA";
+    repo = "NVIDIAOpticalFlowSDK";
+    rev = "79c6cee80a2df9a196f20afd6b598a9810964c32";
+    sha256 = "1y6igwv75v1ynqm7j6la3ky0f15mgnj1jyyak82yvhcsx1aax0a1";
+  };
+
+  # # We only need the header files. The library files are
+  # # in the nvidia_x11 driver.
+  installPhase = ''
+    mkdir -p $out/include
+    cp -R * $out/include
+  '';
+
+  meta = with stdenv.lib; {
+    description = "Nvidia optical flow headers for computing the relative motion of pixels between images";
+    homepage = "https://developer.nvidia.com/opticalflow-sdk";
+    license = licenses.bsd3; # applies to the header files only
+    platforms = platforms.all;
+  };
+}
+

--- a/pkgs/development/libraries/opencv/4.x.nix
+++ b/pkgs/development/libraries/opencv/4.x.nix
@@ -15,7 +15,7 @@
 , enableContrib   ? true
 
 , enableCuda      ? (config.cudaSupport or false) &&
-                    stdenv.hostPlatform.isx86_64, cudatoolkit
+                    stdenv.hostPlatform.isx86_64, cudatoolkit, nvidia-optical-flow-sdk
 
 , enableUnfree    ? false
 , enableIpp       ? false
@@ -166,6 +166,7 @@ stdenv.mkDerivation {
   # Also, work around https://github.com/NixOS/nixpkgs/issues/26304 with
   # what appears to be some stray headers in dnn/misc/tensorflow
   # in contrib when generating the Python bindings:
+  patches = lib.optional enableCuda ./cuda_opt_flow.patch;
   postPatch = ''
     sed -i '/Add these standard paths to the search paths for FIND_LIBRARY/,/^\s*$/{d}' CMakeLists.txt
     sed -i -e 's|if len(decls) == 0:|if len(decls) == 0 or "opencv2/" not in hdr:|' ./modules/python/src2/gen2.py
@@ -213,7 +214,7 @@ stdenv.mkDerivation {
     # tesseract & leptonica.
     ++ lib.optionals enableTesseract [ tesseract leptonica ]
     ++ lib.optional enableTbb tbb
-    ++ lib.optional enableCuda cudatoolkit
+    ++ lib.optionals enableCuda [ cudatoolkit nvidia-optical-flow-sdk ]
     ++ lib.optionals stdenv.isDarwin [ bzip2 AVFoundation Cocoa VideoDecodeAcceleration ]
     ++ lib.optionals enableDocs [ doxygen graphviz-nox ];
 
@@ -249,6 +250,7 @@ stdenv.mkDerivation {
     "-DCUDA_FAST_MATH=ON"
     "-DCUDA_HOST_COMPILER=${cudatoolkit.cc}/bin/cc"
     "-DCUDA_NVCC_FLAGS=--expt-relaxed-constexpr"
+    "-DNVIDIA_OPTICAL_FLOW_1_0_HEADERS_PATH=${nvidia-optical-flow-sdk}"
   ] ++ lib.optionals stdenv.isDarwin [
     "-DWITH_OPENCL=OFF"
     "-DWITH_LAPACK=OFF"

--- a/pkgs/development/libraries/opencv/cuda_opt_flow.patch
+++ b/pkgs/development/libraries/opencv/cuda_opt_flow.patch
@@ -1,0 +1,26 @@
+diff --git a/opencv_contrib/cudaoptflow/CMakeLists.txt b/opencv_contrib/cudaoptflow/CMakeLists.txt
+index e5b823ab4a..a728060d0b 100644
+--- a/opencv_contrib/cudaoptflow/CMakeLists.txt
++++ b/opencv_contrib/cudaoptflow/CMakeLists.txt
+@@ -11,18 +11,6 @@ ocv_define_module(cudaoptflow opencv_video opencv_optflow opencv_cudaarithm open
+ set(NVIDIA_OPTICAL_FLOW_1_0_HEADERS_COMMIT "79c6cee80a2df9a196f20afd6b598a9810964c32")
+ set(NVIDIA_OPTICAL_FLOW_1_0_HEADERS_MD5 "ca5acedee6cb45d0ec610a6732de5c15")
+ set(NVIDIA_OPTICAL_FLOW_1_0_HEADERS_PATH "${OpenCV_BINARY_DIR}/3rdparty/NVIDIAOpticalFlowSDK_1_0_Headers")
+-ocv_download(FILENAME "${NVIDIA_OPTICAL_FLOW_1_0_HEADERS_COMMIT}.zip"
+-               HASH ${NVIDIA_OPTICAL_FLOW_1_0_HEADERS_MD5}
+-               URL
+-                 "https://github.com/NVIDIA/NVIDIAOpticalFlowSDK/archive/"
+-               DESTINATION_DIR "${NVIDIA_OPTICAL_FLOW_1_0_HEADERS_PATH}"
+-               STATUS NVIDIA_OPTICAL_FLOW_1_0_HEADERS_DOWNLOAD_SUCCESS
+-               ID "NVIDIA_OPTICAL_FLOW"
+-               UNPACK RELATIVE_URL)
+ 
+-if(NOT NVIDIA_OPTICAL_FLOW_1_0_HEADERS_DOWNLOAD_SUCCESS)
+-  message(STATUS "Failed to download NVIDIA_Optical_Flow_1_0 Headers")
+-else()
+-  add_definitions(-DHAVE_NVIDIA_OPTFLOW=1)
+-  ocv_include_directories(SYSTEM "${NVIDIA_OPTICAL_FLOW_1_0_HEADERS_PATH}/NVIDIAOpticalFlowSDK-${NVIDIA_OPTICAL_FLOW_1_0_HEADERS_COMMIT}")
+-endif()
+\ No newline at end of file
++add_definitions(-DHAVE_NVIDIA_OPTFLOW=1)
++ocv_include_directories(SYSTEM "${NVIDIA_OPTICAL_FLOW_1_0_HEADERS_PATH}")

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -13585,6 +13585,8 @@ in
 
   nvidia-video-sdk = callPackage ../development/libraries/nvidia-video-sdk { };
 
+  nvidia-optical-flow-sdk = callPackage ../development/libraries/nvidia-optical-flow-sdk { };
+
   nvtop = callPackage ../tools/system/nvtop {
     nvidia_x11 = linuxPackages.nvidia_x11.override { libsOnly = true; };
   };

